### PR TITLE
Add emscripten_promise_await based on ASYNCIFY

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

**Comment by sbc100** [comment]
Still needs more tests, but sending out now so you can check out the general direction.

File: src/library_promise.js [hunk 1]
```diff
@@ -260,5 +259,32 @@ mergeInto(LibraryManager.library, {
...
+#if ASYNCIFY
+#if RUNTIME_DEBUG
+    dbg(`emscripten_promise_wait: ${id}`);
+#endif
+    Asyncify.handleSleep((wakeUp) => {
+      getPromise(id).then((value) => {
+#if ASSERTIONS
+        assert(typeof value === "number", `native promises can only resolve to numbers (${value} ${typeof value})`);
+#endif
+        err('emscripten_promise_wait: promise fulfilled');
```
[the remark this answers is not in the document]
Comment (sbc100) [reply src/library_promise.js:276]: Removed.

**Comment by tlively** [comment]
It looks like the major difference between this and #18603 is that this PR returns to the user when the promise is rejected whereas the other PR never resumes when the promise is rejected. This isn't a huge deal because the user can always chain another promise to handle rejection before waiting.

@sbc100, do you have a strong preference between the approaches?

**Comment by sbc100** [comment]
> It looks like the major difference between this and #18603 is that this PR returns to the user when the promise is rejected whereas the other PR never resumes when the promise is rejected. This isn't a huge deal because the user can always chain another promise to handle rejection before waiting.
> 
> @sbc100, do you have a strong preference between the approaches?

As a user of promise API I think I would almost always be interested in rejections, no?   Why not build it in?

**Comment by sbc100** [comment]
> > It looks like the major difference between this and #18603 is that this PR returns to the user when the promise is rejected whereas the other PR never resumes when the promise is rejected. This isn't a huge deal because the user can always chain another promise to handle rejection before waiting.
> > @sbc100, do you have a strong preference between the approaches?
> 
> As a user of promise API I think I would almost always be interested in rejections, no? Why not build it in?

BTW, I'm happy to land your version, but it does seem good to be able to handle rejections.

**Comment by tlively** [comment]
Let's land this version. I think you're right about rejections. Having handling built into the waiting API is simpler than having to do another `emscripten_promise_then` before waiting just to handle rejections separately.

Can we rename the function to `emscripten_promise_await` to match the similar JS keyword?

**Comment by sbc100** [comment]
> Let's land this version. I think you're right about rejections. Having handling built into the waiting API is simpler than having to do another `emscripten_promise_then` before waiting just to handle rejections separately.
> 
> Can we rename the function to `emscripten_promise_await` to match the similar JS keyword?

Sure, if you think its a good idea to invoke that meaning.  `await` is kind of different to blocking though.. which is why though it best to avoid it.

**Comment by sbc100** [comment]
> > Let's land this version. I think you're right about rejections. Having handling built into the waiting API is simpler than having to do another `emscripten_promise_then` before waiting just to handle rejections separately.
> > Can we rename the function to `emscripten_promise_await` to match the similar JS keyword?
> 
> Sure, if you think its a good idea to evoke that meaning. `await` is kind of different to blocking though.. which is why I thought it best to avoid it (maybe?)

I guess I could do either way on this... i wonder what others think.  @brendandahl @dschuff ?

**Comment by dschuff** [comment]
I guess `wait` lines up with other C APIs like `pthread_cond_wait` (and `futex_wait` and the wasm atomic wait instruction, although those aren't really exposed to users). The usage of this compared to JS `await` does seem similar though? Control does not return past the await until the event loop has turned in either case, right?

**Comment by tlively** [comment]
Yeah, this isn't entirely analogous to JS `await` because there are no async functions (i.e. the C function is not going to magically return a promise because it contains a call to `emscripten_promise_await`), but apart from that, this does do almost exactly the same thing as JS `await` and I think it's useful to explicitly draw attention to that similarity.

**Comment by sbc100** [comment]
Renamed to `await`.. added some comments from the other PR and added more testing.

**Comment by brendandahl** [comment]
`await` seems good, that's what we use in [embind's val](https://github.com/emscripten-core/emscripten/blob/b182182747e8a5525ff3282953fb814c8c0c9266/system/include/emscripten/val.h#L596).

On another thought, what's the use case for using promise + asyncify/jspi? Wouldn't it be easier to use EM_ASYNC_JS or make the import an async function so all the promise code is not observable from c++? I see this let's the user handle rejection values, but I'm not sure what else this gives us.

**Comment by sbc100** [comment]
> `await` seems good, that's what we use in [embind's val](https://github.com/emscripten-core/emscripten/blob/b182182747e8a5525ff3282953fb814c8c0c9266/system/include/emscripten/val.h#L596).
> 
> On another thought, what's the use case for using promise + asyncify/jspi? Wouldn't it be easier to use EM_ASYNC_JS or make the import an async function so all the promise code is not observable from c++? I see this let's the user handle rejection values, but I'm not sure what else this gives us.

The promise API is all about exposing promises a first class object to C/C++ code.   Assuming we agree this is a useful thing to have it seems like the ability to wait on a promise that somebody passed you is pretty useful, no?   

Or are you arguing that the whole concept of first class promises in C/C++ is not needed?

**Comment by sbc100** [comment]
Improved test code .. PTAL

**Comment by brendandahl** [comment]
More the latter, I'm thinking about what the need for promises is when we have ASYNCIFY/JSPI which abstracts them away from c++. I'm not against the idea, more trying to think of concrete uses. For example, if the user was doing a fetch to get some data:

With promises:
```c
EM_JS_DEPS(deps, "$makePromise");

EM_JS(em_promise_t, fetch, (), {
  var info = makePromise();
  fetch("http://example.com/test.json").then((response) => {
    return response.json();
  }).then((json) => {
    info.resolve(json['answer']);
  });
  return info.id;
});


int main() {
  em_promise_t p = fetch();

  em_settled_result_t res = emscripten_promise_await(p);
  printf(".. done wait: %d %ld\n", res.result, (intptr_t)res.value);
  return 0;
}
```
vs JSPI alone
```c
EM_ASYNC_JS(int, fetch, (), {
  let response = await fetch("http://example.com/test.json");
  let json = await response.json();
  return json['answer'];
});


int main() {
  int answer = fetch();
  return 0;
}
```

It seems like promise+JSPI would be useful if things are structured to be driven more from the JS side.  e.g. if you want your c++ to do some work when a promise is ready e.g.
```js
let info = makePromise();
/* ... some async thing that will resolve the above */
let answer = await Module.doSomeWorkWhenReady(info.id);
```

**Comment by tlively** [comment]
The value I see in this API is that we have plenty of async native APIs in Emscripten (e.g. proxying, dlopen, initializing WasmFS backends, etc) and it's nicer if they are built on promises rather than callbacks for all the same reasons promises are nicer than callbacks in JS. It would be awkward if the user was forced to use JS to manipulate the promises returned from these APIs, so we provide a native API for manipulating them instead.

We also wouldn't want to make these APIs "blocking" by having them use JSPI as an implementation detail because that would take control away from users. For example, consider the case when a user wants to asynchronously proxy many tasks at once, then wait for them all to be completed. If we only provided blocking versions of the proxying API, the tasks could only be proxied in serial. But with the promise API and JSPI integration from this PR, the user can instead do `emscripten_promise_all` and use JSPI to await that while all the proxied tasks are performed in parallel.

**Comment by sbc100** [comment]
I think we are outside the scope of this PR at this point, but I think there are some good reasons to sometimes want async/await type semantics and sometimes what to deal with first class promises so they can be composed in various ways rather than just always blocking on one at a time (I saw @tlively already mentioned some).

File: system/include/emscripten/promise.h [hunk 2]
```diff
@@ -135,6 +135,19 @@ __attribute__((warn_unused_result)) em_promise_t emscripten_promise_any(
...
 emscripten_promise_race(em_promise_t* promises, size_t num_promises);
 
+// Suspend the current Wasm execution context until the given promise has been
+// settled.
+//
+// Since the stack is not unwound while Wasm execution is suspended, it is
+// safe to pass pointers to the stack to asynchronous work that is waited on
+// with this function.  If an exception occurs while the promise handler is
+// running then the suspended Wasm will never be resumed and this function
+// will never return.
```
Comment (tlively) [inline system/include/emscripten/promise.h:145]: Looks like this part of the comment needs to be updated for this new approach.

**Review by tlively — APPROVED** [review]
Test LGTM.

File: system/include/emscripten/promise.h (hunk 2 above)
Comment (sbc100) [reply system/include/emscripten/promise.h:145]: Do exceptions automatically turn into promise rejections?

**Pull request merged**